### PR TITLE
add docstrings for loadFromFile

### PR DESCRIPTION
**Pull request by madhur-tandon**
cc: @fpliger

### CODE REVIEW

File: pyscriptjs/src/pyodide.ts [hunk 1]
```diff
@@ -97,6 +97,32 @@ export class PyodideRuntime extends Runtime {
...
     }
 
+    /**
+     *
+     * @param path : the path in the filesystem
+     * @param fetch_path : the path to be fetched
+     *
+     * For a given resource with `fetch_path` as `http://dummy.com/hi.py`
+     * and the location of `path` in the FS to be `a/b/c/foo.py`
+     *
```
Comment (FabioRosado) [inline pyscriptjs/src/pyodide.ts:107]: Do you think it makes sense adding an example in the format: `await loadFromFile("http://dummy.com/hi.py", "a/b/c/foo.py")`?

Just wondering if a visual cue may be a good idea

**Review by FabioRosado — APPROVED** [review]
Thanks! Small comment but is more of a question haha

File: pyscriptjs/src/pyodide.ts [hunk 2]
```diff
@@ -97,6 +97,32 @@ export class PyodideRuntime extends Runtime {
         }
     }
 
+    /**
+     *
+     * @param path : the path in the filesystem
+     * @param fetch_path : the path to be fetched
+     *
+     * For a given resource with `fetch_path` as `http://dummy.com/hi.py`
+     * and the location of `path` in the FS to be `a/b/c/foo.py`
```
Comment (fpliger) [inline pyscriptjs/src/pyodide.ts:106]: This reads a bit confusing to me. A good best practice is to start with a good summary sentence of what the function does. Something like  "Given a file available at a `fetch_path` URL (i.e. `http://dummy.com/hi.py`), downloads the such file and saves to the `path` on the FS."

File: pyscriptjs/src/pyodide.ts [hunk 3]
```diff
@@ -97,6 +97,32 @@ export class PyodideRuntime extends Runtime {
...
+     *
+     * Iteratively analyses the paths:
+     *  - `a`
+     *  - `a/b`
+     *  - `a/b/c`
+     *
+     * The analysis returns if the path exists and if it's parent directory exists
+     * If the path itself doesn't exist, we create it
+     *
+     * Due to the manner in which we proceed, the parent will ALWAYS exist.
```
Comment (fpliger) [inline pyscriptjs/src/pyodide.ts:116]: Similarly to the comment above, I think this over complicates the message.... As a reader, I think I'd be easier for me to understand if we just said something like: "Given a nested `path` such as `a/b/c`, it'll recursively check if each part of the path exists and create it if not."

File: pyscriptjs/src/pyodide.ts [hunk 4]
```diff
@@ -97,6 +97,32 @@ export class PyodideRuntime extends Runtime {
...
+     *
+     * Thus, the iteration proceeds in the following manner:
+     *
+     * - `a` doesn't exist but it's parent i.e. `root` exists --> create `a`
+     * - `a/b` doesn't exist but it's parent i.e. `a` exists --> create `a/b`
+     * - `a/b/c` doesn't exist but it's parent i.e. `a/b` exists --> create `a/b/c`
```
Comment (fpliger) [inline pyscriptjs/src/pyodide.ts:122]: I think it's fine to keep this part

File: pyscriptjs/src/pyodide.ts (hunk 3 above)
Comment (madhur-tandon) [reply pyscriptjs/src/pyodide.ts:116]: I wanted to give a detailed example, but I can make it shorter of course 😅

File: pyscriptjs/src/pyodide.ts [hunk 5]
```diff
@@ -97,6 +97,32 @@ export class PyodideRuntime extends Runtime {
...
+     *
+     * Finally, write content of `http://dummy.com/hi.py` to `a/b/c/foo.py`
```
Comment (fpliger) [inline pyscriptjs/src/pyodide.ts:124]: What about folders? Can I pass a path like `a/b/c` and expect the file to be written as `a/b/c/hi.py`or we always expect a full file path?

Any other edge case worth mentioning?

Any know failure path that we should document?

File: pyscriptjs/src/pyodide.ts [hunk 6]
```diff
@@ -97,6 +97,32 @@ export class PyodideRuntime extends Runtime {
...
+     */
```
Comment (fpliger) [inline pyscriptjs/src/pyodide.ts:125]: Can you also add comments? As a consumer of the code, I'd like a comment that saw:

* what does `const { exists, parentExists } = this.interpreter.FS.analyzePath(eachPath);` and the lines `130 to 138` are doing (basically, recursively create the path on the FS)
* `const response = await robustFetch(fetch_path);` is the response being checked? What happens if it fails to fetch?
* anything in particular that should be a comment on how `const stream = this.interpreter.FS.open(pathArr.join('/'), 'w');` works?

File: pyscriptjs/src/pyodide.ts (hunk 2 above)
Comment (madhur-tandon) [reply pyscriptjs/src/pyodide.ts:106]: Thanks, changed a bit!

File: pyscriptjs/src/pyodide.ts (hunk 3 above)
Comment (madhur-tandon) [reply pyscriptjs/src/pyodide.ts:116]: Done, although the approach is `iterative` here and not recursive.

File: pyscriptjs/src/pyodide.ts (hunk 5 above)
Comment (madhur-tandon) [reply pyscriptjs/src/pyodide.ts:124]: Thanks, I highlighted this in a NOTE at the end.
`a/b/c` is not valid basically, we need a full file path.

Any other edge cases don't come to my mind right now.

File: pyscriptjs/src/pyodide.ts (hunk 6 above)
Comment (madhur-tandon) [reply pyscriptjs/src/pyodide.ts:125]: done

File: pyscriptjs/src/pyodide.ts (hunk 1 above)
Comment (madhur-tandon) [reply pyscriptjs/src/pyodide.ts:108]: Thanks, added!

**Comment by madhur-tandon** [comment]
> some of them should be addressed before we merge

Done all, should be ready I guess

**Review by fpliger — APPROVED** [review]
Overall lgtm, thank you @madhur-tandon !

**Comment by antocuni** [comment]
I'm too late in the conversation, but at some point we should rename this function.
It's not really "loading" anything: what it does is to fetch an URL and save it to the the FS, so maybe it should be called `downloadURL`, `fetchAndSave` or something like this

**Comment by antocuni** [comment]
> I'm too late in the conversation, but at some point we should rename this function:
- It's not really "loading" anything
- it's not "from file" but "from URL"

What it does is to fetch an URL and save it to the the FS, so maybe it should be called `downloadURL`, `fetchAndSave` or something like this

**Comment by fpliger** [comment]
Great point. +1 on `fetchAndSave`

**Pull request merged**